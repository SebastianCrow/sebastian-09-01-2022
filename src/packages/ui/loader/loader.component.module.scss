@import '../../../styles/variables';
@import '../../../styles/themes/dark';

.container {
  padding: 4 * $spacing;

  .loader {
    color: $color-action !important;
  }
}
