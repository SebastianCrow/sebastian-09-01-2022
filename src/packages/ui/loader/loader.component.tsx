import React, { FunctionComponent } from 'react';
import { CircularProgress } from '@mui/material';
import styles from './loader.component.module.scss';

export const Loader: FunctionComponent = () => {
  return (
    <div className={styles.container}>
      <CircularProgress className={styles.loader} />
    </div>
  );
};
