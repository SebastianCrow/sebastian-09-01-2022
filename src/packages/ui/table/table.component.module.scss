.row {
  // TODO: Explore this method
  //background-image: url('../../../../resources/orderBook/highlightAsks.png');
  //background-repeat: no-repeat;
}
