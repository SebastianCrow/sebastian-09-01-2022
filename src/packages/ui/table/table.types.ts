import { ReactNode } from 'react';

export interface ColumnInfo {
  key: string;
  title: ReactNode;
  textAlignment?: 'left' | 'right';
}

export interface RowInfo {
  id: string; // TODO: How to make it better?
  cells: Record<string, CellInfo>; // TODO: Type safety with generics?
}

interface CellInfo {
  value: ReactNode;
  textAlignment?: TextAlignment;
}

export type TextAlignment = 'left' | 'right';
