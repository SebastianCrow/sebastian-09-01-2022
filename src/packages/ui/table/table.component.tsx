import React, { FunctionComponent, useMemo } from 'react';
import cns from 'classnames';
import { ColumnInfo, RowInfo, TableInfo } from './table.types';
import { TEXT_ALIGNMENT } from './table.defaults';
import styles from './table.component.module.scss';
import { Loader } from '../loader/loader.component';

export interface TableProps {
  columns: ColumnInfo[];
  data?: RowInfo[];
  options?: TableInfo;
}

export const Table: FunctionComponent<TableProps> = ({
  columns,
  data,
  options: { headerVisible = true } = {},
}) => {
  if (columns.length === 0) {
    throw new Error('There needs to be at least one column definition');
  }
  const colWidth = useMemo(() => `${100 / columns.length}%`, [columns.length]);
  if (!data) {
    return (
      <div className={styles.loadingOverlay}>
        <Loader />
      </div>
    );
  }
  return (
    <div>
      {headerVisible && (
        <div className={styles.rowHeader}>
          {columns.map(({ key, title, textAlignment = TEXT_ALIGNMENT }) => (
            <div
              key={key}
              className={cns(styles.cell, {
                [styles.alignRight]: textAlignment === 'right',
              })}
              style={{ width: colWidth }}
            >
              {title}
            </div>
          ))}
        </div>
      )}
      {data.map(({ id, rowClass, rowStyle, cells }, index) => (
        <div
          key={`${id}_${index}`}
          className={cns(styles.rowData, rowClass)}
          style={rowStyle}
        >
          {columns.map(({ key }) => {
            const cell = cells[key];
            if (!cell) {
              throw new Error(
                `Cell info for data id=[${id}] and column key=[${key}] not found`
              );
            }
            const { value, textAlignment = TEXT_ALIGNMENT } = cell;
            return (
              <div
                key={key}
                className={cns(styles.cell, {
                  [styles.alignRight]: textAlignment === 'right',
                })}
                style={{ width: colWidth }}
              >
                {/* TODO: Wrap value with <Text /> if it's not a custom component */}
                {value}
              </div>
            );
          })}
        </div>
      ))}
    </div>
  );
};
