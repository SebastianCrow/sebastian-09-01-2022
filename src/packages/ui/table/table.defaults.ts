import { TextAlignment } from './table.types';

export const TEXT_ALIGNMENT: TextAlignment = 'right';
