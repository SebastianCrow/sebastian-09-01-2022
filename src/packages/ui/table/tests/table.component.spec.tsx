import React from 'react';
import { Table } from '../table.component';
import { create } from 'react-test-renderer';
import { Loader } from '../../loader/loader.component';

describe('table.component', () => {
  test('renders loader without data', () => {
    const renderer = create(
      <Table columns={[{ key: 'test', title: 'Key' }]} />
    );
    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(Loader)).toHaveLength(1);
  });

  test('renders column', () => {
    const renderer = create(
      <Table
        columns={[{ key: 'testColumn', title: 'Test Column' }]}
        data={[
          { id: 'testRow', cells: { testColumn: { value: 'Test Cell' } } },
        ]}
      />
    );
    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(Loader)).toHaveLength(0);
  });

  // TODO: More specific tests for values in the table
});
