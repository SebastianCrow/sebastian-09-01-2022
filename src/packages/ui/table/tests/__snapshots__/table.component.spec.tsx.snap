// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table.component renders column 1`] = `
<div>
  <div
    className="rowHeader"
  >
    <div
      className="cell alignRight"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      Test Column
    </div>
  </div>
  <div
    className="rowData"
  >
    <div
      className="cell alignRight"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      Test Cell
    </div>
  </div>
</div>
`;

exports[`table.component renders loader without data 1`] = `
<div
  className="loadingOverlay"
>
  <div
    className="container"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>
</div>
`;
